### DISCUSSION TITLE
Cargo options to exclude CMAKE?

### DISCUSSION THREAD

**Discussion by rimutaka** [GitoxideLabs/gitoxide#122] [Q&A]
Is there an option to swap `zlib-ng` for a native Rust implementation?

I see there is a whole lot of different options, but none maps directly to disabling `zlib-ng`. 

I got this on `cargo run` after cloning the repo:
```
error: failed to run custom build command for `libz-sys v1.1.3`

Caused by:
  process didn't exit successfully: `/home/mx/projects/stm/test-repos/rust/gitoxide/target/debug/build/libz-sys-54f74038388b2a05/build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-env-changed=LIBZ_SYS_STATIC
  cargo:rerun-if-changed=build.rs
  running: "cmake" "/home/mx/.cargo/registry/src/github.com-1ecc6299db9ec823/libz-sys-1.1.3/src/zlib-ng" "-DBUILD_SHARED_LIBS=OFF" "-DZLIB_COMPAT=ON" "-DWITH_GZFILEOP=ON" "-DCMAKE_INSTALL_PREFIX=/home/mx/projects/stm/test-repos/rust/gitoxide/target/debug/build/libz-sys-28e392133e6cd8e2/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_C_COMPILER=/usr/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_CXX_COMPILER=/usr/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_ASM_COMPILER=/usr/bin/cc" "-DCMAKE_BUILD_TYPE=Debug"
```

I only really need `git-log` at this stage.

**Comment by edward-shen** [comment]
Just a heads-up, we decided on using `zlib-ng` over a native rust implementation due to the large performance benefits; you can see the discussion in [#47](https://github.com/Byron/gitoxide/discussions/47#discussioncomment-408740). You might not need it, but it might be worthwhile to try to get it to work depending on your use case.

**Comment by Byron** [comment]
I am glad you ask :).

```
cargo build --no-default-features --features small
```

The line above will do the trick. Currently the `small` feature is configured like this…

```toml
small = ["lean-cli", "git-features/sha1", "git-features/zlib-rust-backend"]
```

…which also means that multi-threading will be disabled and everything will be using a Rust implementation.

It's totally possible to re-enable parallelism and get something like a `pure-rust` target to get all the features and best-possible performance while avoiding C implementations.

This kind of configuration is meant to be done by applications using `gitoxide`, similar to how the top-level `Cargo.toml` file already does it.

All features are [explained here](https://github.com/Byron/gitoxide/blob/4b63d8a8709a2674d287879c4d6538a74cd7869b/cargo-features.md#L1-L0).

> I only really need git-log at this stage.

There is no `git-log` like capability in the `gix` or `gixp` commands right now.

In case you are interested in creating your own application, I recommend pulling in `git-repository` for a reasonably high-level abstraction that is quite WIP at the moment, but capable of doing git-log like things for sure.  Configure it [like so](https://github.com/Byron/gitoxide/blob/352e4689959dee169f08b8fbd7be3c1f234202b8/gitoxide-core/Cargo.toml#L29-L30).

Lastly, you can configure everything else in `git-features`. You can refer to it similar to [to this](https://github.com/Byron/gitoxide/blob/0357dc9d1f390b101868421756f459158fc84167/Cargo.toml#L69) and enable [additional features](https://github.com/Byron/gitoxide/blob/4b63d8a8709a2674d287879c4d6538a74cd7869b/cargo-features.md#L87-L86) as you see fit.

If you run into trouble or have any additional questions, please let me know as this would be considered a bug. The idea is that applications have a lot of control over what goes into it.

**Comment by rimutaka** [comment]
OK, thank you guys! It compiles now.

Can you share why the performance is your # 1 metric? What is the use case?

**Comment by colbyn** [comment]
The problem with C libs is that it severely complicates cross compilation, for instance compiling this library for IOS. 

Personally I think there should always be an option to opt-out of C deps if a native rust version is available.
